simSPI/linear_simulator/projector_utils.py

```diff
@@ -0,0 +1,71 @@
+"""Class to generate tomographic projection."""
+
+import torch
+from pytorch3d.transforms import Rotate
+
+
+class Projector(torch.nn.Module):
+    """Class to generate tomographic projection.
+
+    Written by J.N. Martel, Y. S. G. Nashed, and Harshit Gupta.
+
+    Parameters
+    ----------
+    config: class
+        Class containing parameters of the Projector
+
+    """
+
+    def __init__(self, config):
+        super(Projector, self).__init__()
+
+        self.config = config
+        self.vol = torch.rand([self.config.side_len] * 3, dtype=torch.float32)
+        lincoords = torch.linspace(-1.0, 1.0, self.config.side_len)
+        [X, Y, Z] = torch.meshgrid([lincoords, lincoords, lincoords])
+        coords = torch.stack([Y, X, Z], dim=-1)
+        self.register_buffer("vol_coords", coords.reshape(-1, 3))
+
+    def forward(self, rot_params, proj_axis=-1):
+        """Output the tomographic projection of the volume.
+
+        First rotate the volume and then sum it along an axis.
+        The volume is assumed to be square. The output image
+        follows (batch x channel x height x width) convention of pytorch.
+        Therefore, a dummy channel dimension is added at the end to projection.
+
+        Parameters
+        ----------
+        rot_params: dict of type str to {tensor}
+            contains rotation matrix "rotmat" (batch_size x 3 x 3) that is
+            used to rotate the volume
+        proj_axis: int
+            index along which summation is done of the rotated volume
+
+        Returns
+        -------
+        projection: tensor
+            Tensor containing tomographic projection
+            (batch_size x 1 x sidelen x sidelen)
+        """
+        rotmat = rot_params["rotmat"]
+        batch_sz = rotmat.shape[0]
```

Comment:
NIT: len(rotmat)

Reply:
I prefer .shape[0] because it is much more flexible as it can return shape of any dimension and not just the first dimension. Therefore, it is more future proof.